### PR TITLE
CORE-4998 Cleanup flow mapper in-memory state

### PR DESCRIPTION
**Pull request by lankydan**
Clean up flow mapper in-memory state of flow starts and sessions.

When a flow finishes of fails, clean up its state.

When a session becomes `CLOSED` or `ERROR`ed, clean up its state.

State is cleaned up by sending a `ScheduleCleanup` record to the flow
mapper, which cleans up the in-memory state after a period of time.

`FlowCheckpointImpl` now _soft deletes_ the Avro `Checkpoint` so that
sessions can still be accessed after marking the `Checkpoint` for
deletion. This is needed if a session goes to `CLOSED` and the flow
finishes all off of a single event, otherwise the `ScheduleCleanup`
record cannot be sent for the session.

This commit also changes the flow mapper's key solely to a `String`. The
string still represents a `FlowKey` (for flow starts) or a session id
(for a session). The change calls `toString` on `FlowKey` rather than
passing in the `FlowKey` directly and letting the state and event
pattern do the conversion to a `String`. This was required to fix a bug
where the partition a state/event was read from was not the same
partition the new state was written to, which indicated a change in the
generated hash of the key. Calling `toString` made the output
consistent and preventing changing the partition.

### CODE REVIEW

File: components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/CloseSessionsAcceptanceTest.kt [hunk 1]
```diff
@@ -377,59 +402,13 @@ class CloseSessionsAcceptanceTest : FlowServiceTestBase() {
...
-     * Given two sessions where:
-     *
-     * - `SESSION_ID_1` has already received a session close event.
-     * - `close` is called.
-     * - `SESSION_ID_2` receives a session close event.
-     *
-     * The flow does not resume and sends `SESSION_ID_2` a session ack.
-     */
-    @Test
-    fun `Given two sessions where one has already received a session close event calling close and then receiving a session close event for the other session does not resume the flow and sends a session ack`() {
```
Comment (JamesHR3) [inline components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/CloseSessionsAcceptanceTest.kt:399]: Why did you remove this test out of interest?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/FlowFailedRequestHandler.kt [hunk 2]
```diff
@@ -31,18 +34,23 @@ class FlowFailedRequestHandler @Activate constructor(
...
 
         val status = flowMessageFactory.createFlowFailedStatusMessage(
-            context.checkpoint,
+            checkpoint,
             FLOW_FAILED,
             request.exception.message ?: request.exception.javaClass.name
         )
 
-        val record = flowRecordFactory.createFlowStatusRecord(status)
+        val expiryTime = Instant.now().plusMillis(1.minutes.toMillis()).toEpochMilli() // TODO Should be configurable?
```
Comment (JamesHR3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/FlowFailedRequestHandler.kt:47]: It probably should be configurable - could you make a tech debt item for it?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/state/impl/FlowCheckpointImpl.kt [hunk 3]
```diff
@@ -102,7 +104,7 @@ class FlowCheckpointImpl(
         get() = sessionMap.values.toList()
 
     override val doesExist: Boolean
-        get() = nullableCheckpoint != null
+        get() = nullableCheckpoint != null && !deleted
```
Comment (JamesHR3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/state/impl/FlowCheckpointImpl.kt:108]: Do you need to check both here? Feels like we should just switch to using the `deleted` flag everywhere in this class to establish whether the object is about to be deleted - particularly as  I don't think you'd actually null out the checkpoint any more.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/state/impl/FlowCheckpointImpl.kt [hunk 4]
```diff
@@ -149,18 +151,21 @@ class FlowCheckpointImpl(
     }
 
     override fun putSessionState(sessionState: SessionState) {
+        checkFlowNotDeleted()
         sessionMap[sessionState.sessionId] = sessionState
     }
 
     override fun markDeleted() {
-        nullableCheckpoint = null
+        checkFlowNotDeleted()
```
Comment (JamesHR3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/state/impl/FlowCheckpointImpl.kt]: I think this should be idempotent - doesn't really matter if we mark the checkpoint as deleted twice, only that once we've deleted it we don't try and update it.

**Review by JamesHR3 — COMMENTED** [review]
I know we've discussed this before but I wonder how bad it would be to just always schedule a cleanup for `CLOSING`/`ERRORED` sessions? That would remove the need for the API change you've made here. The mapper should deal with duplicate removal requests already so from that perspective it should be fine. I guess the question is do we ever remove session objects from the checkpoint? If not then we could be scheduling a lot of duplicates, which may or may not be problematic.

Otherwise just a few comments.

**Comment by lankydan** [comment]
> I know we've discussed this before but I wonder how bad it would be to just always schedule a cleanup for `CLOSING`/`ERRORED` sessions? That would remove the need for the API change you've made here. The mapper should deal with duplicate removal requests already so from that perspective it should be fine. I guess the question is do we ever remove session objects from the checkpoint? If not then we could be scheduling a lot of duplicates, which may or may not be problematic.
> 
> Otherwise just a few comments.

Currently, we do not remove `SessionState`s from the checkpoint. Even when a subflow finishes, we just remove the flow stack item and therefore the link to the `SessionState` goes away but the actual object still exists.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/state/impl/FlowCheckpointImpl.kt (hunk 3 above)
Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/state/impl/FlowCheckpointImpl.kt:108]: It needs to remove this way because we use `doesExist` in the `SessionEventHandler` to check if the checkpoint has been created yet. In this case `deleted` would be `false` meaning `doesExist` would return `true` and causing a change in the current behaviour.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/FlowFailedRequestHandler.kt (hunk 2 above)
Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/FlowFailedRequestHandler.kt:47]: Sure.

File: components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/CloseSessionsAcceptanceTest.kt (hunk 1 above)
Comment (lankydan) [reply components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/CloseSessionsAcceptanceTest.kt:399]: I realised it was testing the same behaviour as the test directly below it.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/FlowFailedRequestHandler.kt (hunk 2 above)
Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/FlowFailedRequestHandler.kt:47]: Created - https://r3-cev.atlassian.net/browse/CORE-5077.

File: components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/executor/FlowMapperListener.kt [hunk 5]
```diff
@@ -77,15 +77,15 @@ class FlowMapperListener(
...
-            expiryTime - clock.millis(),
-            TimeUnit.MILLISECONDS
-        )
+        scheduledTasks.computeIfAbsent(eventKey) {
+            executorService.schedule(
+                {
+                    log.debug { "Clearing up mapper state for key $eventKey" }
+                    publisher?.publish(listOf(Record(FLOW_MAPPER_EVENT_TOPIC, eventKey, FlowMapperEvent(ExecuteCleanup()))))
+                },
+                expiryTime - clock.millis(),
```
Comment (simon-johnson-r3) [inline components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/executor/FlowMapperListener.kt:86]: Not directly related to this change, but this makes this function exhibit undefined behaviour if the system UTC time is changed. Is it a requirement to work in absolute time rather than duration? I see it comes from state.expiryTime but I can't track the origin of that down.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt [hunk 6]
```diff
@@ -30,29 +36,44 @@ class FlowGlobalPostProcessorImpl @Activate constructor(
 
         val now = Instant.now()
 
-        val checkpoint = context.checkpoint
+        val outputRecords = getSessionEvents(context, now)
+            .plus(getFlowMapperSessionCleanupEvents(context, now))
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt]: are you calling the operator directly here? below you use +, is it possible to stick to a consistent style?

Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt]: Ye so, I did this because the `+` didn't format very nicely with the amount of lists we were adding together.

It looked like:

```kotlin
getSessionEvents(context, now) + 
    getFlowMapperSessionCleanupEvents(context, now) +
    postProcessRetries(context)
```

I found out when I did this, that you can't put the `+` at the start of the line either, or the compiler gets confused.

Comment (JamesHR3) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt]: I think I prefer the `+` format anyway, that looks clearer to me.

File: components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/executor/FlowMapperListener.kt (hunk 5 above)
Comment (JamesHR3) [reply components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/executor/FlowMapperListener.kt:86]: We have to track an expiry time to cope with the failover case where we might lose this timer to another worker. This does impose a time synchronization constraint though, in that if workers don't agree on a time then you'll get different behaviour depending on which one is maintaining the timer.

**Comment by JamesHR3** [comment]
> > I know we've discussed this before but I wonder how bad it would be to just always schedule a cleanup for `CLOSING`/`ERRORED` sessions? That would remove the need for the API change you've made here. The mapper should deal with duplicate removal requests already so from that perspective it should be fine. I guess the question is do we ever remove session objects from the checkpoint? If not then we could be scheduling a lot of duplicates, which may or may not be problematic.
> > Otherwise just a few comments.
> 
> Currently, we do not remove `SessionState`s from the checkpoint. Even when a subflow finishes, we just remove the flow stack item and therefore the link to the `SessionState` goes away but the actual object still exists.

I'm a bit nervous about introducing flags like this because it feels like the sort of thing where we lose an update somewhere and suddenly the state here is out of sync with the reality on the mapper. Of course, we could have the same problem if we remove the session entirely. Let's go with this for now, but I wouldn't mind revisiting to try and simplify a little if we have time.

**Review by JamesHR3 — DISMISSED** [review]

**Review by JamesHR3 — APPROVED** [review]
